pkg/repo/chartrepo.go

```diff
@@ -249,10 +249,14 @@ func FindChartInAuthAndTLSAndPassRepoURL(repoURL, username, password, chartName,
 	if err != nil {
 		return "", err
 	}
+	// Use a random path name so it can be deleted for cleanup.
+	r.CachePath = helmpath.CachePath(name)
 	idx, err := r.DownloadIndexFile()
 	if err != nil {
 		return "", errors.Wrapf(err, "looks like %q is not a valid chart repository or cannot be reached", repoURL)
 	}
+	// Clean up temp files.
+	defer os.RemoveAll(r.CachePath)
```

Comment:
If we remove line 253, this purges the cache directory on every call to `FindChartInAuthAndTLSAndPassRepoURL`. That is DEFINITELY not what we'd expect, as this is a shared cache directory.

Reply:
I think this only removes the temporary (random) sub-directory of the cache.

Reply:
The current PR changes the cache directory and leaves it changed when the function is done ...

```go
r.CachePath = helmpath.CachePath(name)
```

This also doesn't let an SDK specify the cache location as part of their work. This further intermixes Helm paths into the SDK where an application would want to set their own. Long term we should be on a path to unwind all of this so that the Helm client passes in the cache locations and the SDK takes them from there.

I think putting this here moves the wrong way in addition to not allowing SDK user configurability.

But, there is a legit bug and the fix is headed in the right direction. Instead of removing `r.CachePath` we can remove the temporary file. In `DownloadIndexFile()` we can see where the file name is generated at https://github.com/helm/helm/blob/ea41e790a577e686869023c5687ff8649ec2c16c/pkg/repo/chartrepo.go#L154

So, this PR could do something like...

```go
defer os.RemoveAll(filepath.Join(r.CachePath, helmpath.CacheChartsFile(r.Config.Name)))
```

Note, don't forget to remove the part that sets `r.CachePath`.